### PR TITLE
Prevent anyio.ExceptionGroup in error views under a BaseHTTPMiddleware

### LINKED ISSUE #1255: test_custom_middleware[trio] fails on 1-core-VM
### Checklist

- [X] The bug is reproducible against the latest release and/or `master`.
- [ ] There are no similar issues or pull requests to fix it yet.

### Describe the bug

While working on [reproducible builds](https://reproducible-builds.org/) for [openSUSE](https://en.opensuse.org/openSUSE:Reproducible_Builds), I found that our `python-starlette-0.16.0` package failed 1 test when running in a 1-core-VM (usually this happens due to differences in scheduling/timing)

### To reproduce

maybe run tests as `taskset 1 pytest`

or on Debian or openSUSE run
```
osc checkout openSUSE:Factory/python-starlette && cd $_
osc build --vm-type=kvm --noservice --clean -j1 standard
```

### Expected behavior

tests should pass

### Actual behavior

`test_custom_middleware[trio]` fails

### Debugging material

```
 =================================== FAILURES ===================================
 _________________________ test_custom_middleware[trio] _________________________
 
 test_client_factory = functools.partial(<class 'starlette.testclient.TestClient'>, backend='trio', backend_options={})
 
     def test_custom_middleware(test_client_factory):
         client = test_client_factory(app)
         response = client.get("/")
         assert response.headers["Custom-Header"] == "Example"
     
         with pytest.raises(Exception):
 >           response = client.get("/exc")
 
 tests/middleware/test_base.py:56: 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
 /usr/lib/python3.9/site-packages/requests/sessions.py:555: in get
     return self.request('GET', url, **kwargs)
 starlette/testclient.py:468: in request
     return super().request(
 /usr/lib/python3.9/site-packages/requests/sessions.py:542: in request
     resp = self.send(prep, **send_kwargs)
 /usr/lib/python3.9/site-packages/requests/sessions.py:655: in send
     r = adapter.send(request, **kwargs)
 starlette/testclient.py:266: in send
     raise exc
 starlette/testclient.py:263: in send
     portal.call(self.app, scope, receive, send)
 /usr/lib/python3.9/site-packages/anyio/from_thread.py:229: in call
     return self.start_task_soon(func, *args).result()
 /usr/lib64/python3.9/concurrent/futures/_base.py:445: in result
     return self.__get_result()
 /usr/lib64/python3.9/concurrent/futures/_base.py:390: in __get_result
     raise self._exception
 /usr/lib/python3.9/site-packages/anyio/from_thread.py:176: in _call_func
     retval = await retval
 starlette/applications.py:112: in __call__
     await self.middleware_stack(scope, receive, send)
 starlette/middleware/errors.py:159: in __call__
     await self.app(scope, receive, _send)
 starlette/middleware/base.py:57: in __call__
     task_group.cancel_scope.cancel()
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
 
 self = <anyio._backends._trio.TaskGroup object at 0x7f1d7c50a5b0>
 exc_type = <class 'RuntimeError'>
 exc_val = RuntimeError('No response returned.')
 exc_tb = <traceback object at 0x7f1d7c5294c0>
 
     async def __aexit__(self, exc_type: Optional[Type[BaseException]],
                         exc_val: Optional[BaseException],
                         exc_tb: Optional[TracebackType]) -> Optional[bool]:
         try:
             return await self._nursery_manager.__aexit__(exc_type, exc_val, exc_tb)
         except trio.MultiError as exc:
 >           raise ExceptionGroup(exc.exceptions) from None
 E           anyio._backends._trio.ExceptionGroup: 2 exceptions were raised in the task group:
 E           ----------------------------
 E           Traceback (most recent call last):
 E             File "/home/abuild/rpmbuild/BUILD/starlette-0.16.0/starlette/middleware/base.py", line 30, in coro
 E               await self.app(scope, request.receive, send_stream.send)
 E             File "/home/abuild/rpmbuild/BUILD/starlette-0.16.0/starlette/exceptions.py", line 82, in __call__
 E               raise exc
 E             File "/home/abuild/rpmbuild/BUILD/starlette-0.16.0/starlette/exceptions.py", line 71, in __call__
 E               await self.app(scope, receive, sender)
 E             File "/home/abuild/rpmbuild/BUILD/starlette-0.16.0/starlette/routing.py", line 656, in __call__
 E               await route.handle(scope, receive, send)
 E             File "/home/abuild/rpmbuild/BUILD/starlette-0.16.0/starlette/routing.py", line 259, in handle
 E               await self.app(scope, receive, send)
 E             File "/home/abuild/rpmbuild/BUILD/starlette-0.16.0/starlette/routing.py", line 63, in app
 E               response = await run_in_threadpool(func, request)
 E             File "/home/abuild/rpmbuild/BUILD/starlette-0.16.0/starlette/concurrency.py", line 39, in run_in_threadpool
 E               return await anyio.to_thread.run_sync(func, *args)
 E             File "/usr/lib/python3.9/site-packages/anyio/to_thread.py", line 28, in run_sync
 E               return await get_asynclib().run_sync_in_worker_thread(func, *args, cancellable=cancellable,
 E             File "/usr/lib/python3.9/site-packages/anyio/_backends/_trio.py", line 170, in run_sync_in_worker_thread
 E               return await run_sync(wrapper, cancellable=cancellable, limiter=limiter)
 E             File "/usr/lib/python3.9/site-packages/trio/_threads.py", line 205, in to_thread_run_sync
 E               return await trio.lowlevel.wait_task_rescheduled(abort)
 E             File "/usr/lib/python3.9/site-packages/trio/_core/_traps.py", line 166, in wait_task_rescheduled
 E               return (await _async_yield(WaitTaskRescheduled(abort_func))).unwrap()
 E             File "/usr/lib/python3.9/site-packages/outcome/_sync.py", line 111, in unwrap
 E               raise captured_error
 E             File "/usr/lib/python3.9/site-packages/trio/_threads.py", line 155, in do_release_then_return_result
 E               return result.unwrap()
 E             File "/usr/lib/python3.9/site-packages/outcome/_sync.py", line 111, in unwrap
 E               raise captured_error
 E             File "/usr/lib/python3.9/site-packages/trio/_threads.py", line 168, in worker_fn
 E               ret = sync_fn(*args)
 E             File "/usr/lib/python3.9/site-packages/anyio/_backends/_trio.py", line 168, in wrapper
 E               return func(*args)
 E             File "/home/abuild/rpmbuild/BUILD/starlette-0.16.0/tests/middleware/test_base.py", line 28, in exc
 E               raise Exception()
 E           Exception
 E           ----------------------------
 E           Traceback (most recent call last):
 E             File "/usr/lib/python3.9/site-packages/anyio/streams/memory.py", line 78, in receive
 E               return self.receive_nowait()
 E             File "/usr/lib/python3.9/site-packages/anyio/streams/memory.py", line 73, in receive_nowait
 E               raise WouldBlock
 E           anyio.WouldBlock
 E           
 E           During handling of the above exception, another exception occurred:
 E           
 E           Traceback (most recent call last):
 E             File "/home/abuild/rpmbuild/BUILD/starlette-0.16.0/starlette/middleware/base.py", line 35, in call_next
 E               message = await recv_stream.receive()
 E             File "/usr/lib/python3.9/site-packages/anyio/streams/memory.py", line 98, in receive
 E               raise EndOfStream
 E           anyio.EndOfStream
 E           
 E           During handling of the above exception, another exception occurred:
 E           
 E           Traceback (most recent call last):
 E             File "/home/abuild/rpmbuild/BUILD/starlette-0.16.0/starlette/middleware/base.py", line 55, in __call__
 E               response = await self.dispatch_func(request, call_next)
 E             File "/home/abuild/rpmbuild/BUILD/starlette-0.16.0/tests/middleware/test_base.py", line 12, in dispatch
 E               response = await call_next(request)
 E             File "/home/abuild/rpmbuild/BUILD/starlette-0.16.0/starlette/middleware/base.py", line 37, in call_next
 E               raise RuntimeError("No response returned.")
 E           RuntimeError: No response returned.
 
 /usr/lib/python3.9/site-packages/anyio/_backends/_trio.py:141: ExceptionGroup
 =========================== short test summary info ============================
 SKIPPED [3] tests/conftest.py:14: Trio not supported (yet!)
 =================== 1 failed, 464 passed, 3 skipped in 4.95s ===================
```

### Environment

- OS: Linux = openSUSE Tumbleweed 20210726 
- Python version: 3.9
- Starlette version: 0.16.0

### Additional context

This bug was found while working on [reproducible builds for openSUSE](https://en.opensuse.org/openSUSE:Reproducible_Builds).

### PR DESCRIPTION
**Pull request by florimondmanca**
Closes #1255 

**Motivation**
#1157 switched to using `anyio` for async concurrency operations.

But now when a view raises an exception, say `MyExc`, and a `BaseHTTPMiddleware` is installed, calling the view would make Starlette raise an `anyio.ExceptionGroup`, rather than `MyExc` directly.

Previously, under `asyncio`, we were calling `task.result()` before raising `RuntimeError("No response sent.")`, so only `MyExc` was raised, as expected, see here: https://github.com/encode/starlette/blob/f5a08d019ade528ae3a855f7e52a5500b233cb0a/starlette/middleware/base.py#L45-L46

This PR switches back to that behavior by adding a `checkpoint` (equivalent to `asyncio.sleep(0)` — i.e. "allow switching tasks and checking for cancellation) before raising the "No response sent" exception.

cc @uSpike

### CODE REVIEW

File: tests/middleware/test_base.py [hunk 1]
```diff
@@ -52,8 +52,9 @@ def test_custom_middleware(test_client_factory):
     response = client.get("/")
     assert response.headers["Custom-Header"] == "Example"
 
-    with pytest.raises(Exception):
+    with pytest.raises(Exception) as ctx:
```
Comment (florimondmanca) [inline tests/middleware/test_base.py:55]: When running on `trio`, previously we would be getting an unexpected "Portal is not running" error here, and misinterpreting it as the `Exception()` in the `/exc` view. So I made the latter more precise and made sure we're explicitly checking for it.

**Comment by uSpike** [comment]
I'm not sure I'm in favor of relying on event loop magic to correctly propagate exceptions.

Here's an idea that I think is a bit easier to follow, thoughts?:
```python
diff --git a/starlette/middleware/base.py b/starlette/middleware/base.py
index bf337b8..30ae1ed 100644
--- a/starlette/middleware/base.py
+++ b/starlette/middleware/base.py
@@ -24,21 +24,25 @@ class BaseHTTPMiddleware:
 
         async def call_next(request: Request) -> Response:
             send_stream, recv_stream = anyio.create_memory_object_stream()
+            app_exc: typing.Optional[Exception] = None
 
             async def coro() -> None:
                 async with send_stream:
-                    await self.app(scope, request.receive, send_stream.send)
+                    try:
+                        await self.app(scope, request.receive, send_stream.send)
+                    except Exception as exc:
+                        nonlocal app_exc
+                        app_exc = exc
 
             task_group.start_soon(coro)
 
             try:
                 message = await recv_stream.receive()
             except anyio.EndOfStream:
-                # HACK: give anyio a chance to surface any app exception first,
-                # in order to avoid an `anyio.ExceptionGroup`.
-                # See #1255.
-                await anyio.lowlevel.checkpoint()
-                raise RuntimeError("No response returned.")
+                if app_exc is not None:
+                    raise app_exc
+                else:
+                    raise RuntimeError("No response returned.")
 
             assert message["type"] == "http.response.start"
```

**Comment by simondale00** [comment]
Hey team. Anything I can do to help push this along?

I'm running into this issue as well, which is impacting `Exception` handling. Specifically, adding any user middleware breaks `ServerErrorMiddleware`, as `ServerErrorMiddleware` is not built to catch `ExceptionGroup`. (While one could argue `ServerErrorMiddleware` could be modified to handle exception groups, I think the core issue here is that `BaseHTTPMiddleware` dismisses the possibility that a callable raises an exception and always expects to read a response)

I've tested out the proposal from @uSpike and it works well. Also seems like a pragmatic approach to resolve the issue.

When can we get this merged and released?

**Comment by florimondmanca** [comment]
Looks like this is ready then — happy to have anyone review and release this.

**Review by simondale00 — APPROVED** [review]
Looks good to me. Thanks, @florimondmanca!

**Comment by kissgyorgy** [comment]
Hello, we will be using this fix at our company too, looks good.

**Comment by florimondmanca** [comment]
@Kludex Hi, saw you open #1324. I'd need a ✅ around here to merge. Thanks!

**Review by Kludex — APPROVED** [review]
This solution looks much simpler :+1: 

I was waiting to see if @uSpike wanted to say something, but the code looks pretty straightforward. :smile:

**Comment by florimondmanca** [comment]
Thanks!

**Pull request merged**